### ISSUE TITLE
Feedback about torch.nn.functional.scaled_dot_product_attention

### ISSUE THREAD

**Issue by Seqaeon** [pytorch/pytorch#163026]
There is the following issue on this page: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html

Am I missing something or this implementation doesnt seem to actually mask the attention scores and weights, it just adds the masks to them through the bias term:
`

# Efficient implementation equivalent to the following:
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0,
        is_causal=False, scale=None, enable_gqa=False) -> torch.Tensor:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype, device=query.device)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias = attn_mask + attn_bias

    if enable_gqa:
        key = key.repeat_interleave(query.size(-3)//key.size(-3), -3)
        value = value.repeat_interleave(query.size(-3)//value.size(-3), -3)

    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value
`

**Comment by ArtemF42** [comment]
Hi! You're not missing anything — the function *does* mask attention by adding a bias to the logits **before** the softmax. The `attn_mask` either contains `-inf` in the positions that need to be masked or is converted to that form (if a boolean mask is specified). When added to the attention scores and then passed through softmax, the corresponding values effectively become 0, which is equivalent to proper masking — preventing those positions from contributing to the output.

**Comment by abitpal** [comment]
Conventional idea of masking an attention score means adding -infinity to the attention score to remove the attention score's "influence" during the softmax - this is because e^z approaches 0 as z approaches -inf. In this implementation, if the attention mask is a boolean array, then ```attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))``` sets the attention bias to have the aforementioned behavior where masked out attention scores are set to -infinity. I hope that makes sense.
